cherab/tools/raytransfer/emitters.pyx

```diff
@@ -362,26 +470,34 @@ def dphi(self):
     def dz(self):
         return self._dz
 
-    def emission_function(self, point, direction, spectrum, world, ray, primitive, world_to_primitive, primitive_to_world):
-        iz = int(point.z / self._dz)  # Z-index of grid cell, in which the point is located
-        r = np.sqrt(point.x * point.x + point.y * point.y)  # R coordinates of the points
-        ir = int((r - self._rmin) / self._dr)  # R-index of grid cell, in which the points is located
+    cpdef Spectrum emission_function(self, Point3D point, Vector3D direction, Spectrum spectrum,
+                                     World world, Ray ray, Primitive primitive,
+                                     AffineMatrix3D world_to_primitive, AffineMatrix3D primitive_to_world):
+
+        cdef:
+            int isource, ir, iphi, iz, icell
+            double r, phi
+
+        iz = <int>(point.z / self._dz)  # Z-index of grid cell, in which the point is located
+        r = sqrt(point.x * point.x + point.y * point.y)  # R coordinates of the points
+        ir = <int>((r - self._rmin) / self._dr)  # R-index of grid cell, in which the points is located
         if self.voxel_map.ndim > 2:  # 3D grid
-            phi = (180. / np.pi) * np.arctan2(point.y, point.x)  # phi coordinate of the point (in degrees)
+            phi = (180. / pi) * atan2(point.y, point.x)  # phi coordinate of the point (in degrees)
             if phi < 0:
                 phi += 360.  # moving to [0, 360) interval
             phi = phi % self._period  # moving into the [0, period) sector (periodic emitter)
-            iphi = int(phi / self._dphi)  # phi-index of grid cell, in which the point is located
-            i = self.voxel_map[ir, iphi, iz]  # index of the light source in spectral array
-        else:  # 2D grid (RZ-plane)
-            i = self.voxel_map[ir, iz]  # index of the light source in spectral array
-        if i < 0:  # grid cell is not mapped to any light source
+            iphi = <int>(phi / self._dphi)  # phi-index of grid cell, in which the point is located
+            icell = self._voxel_map.shape[1] * self._voxel_map.shape[2] * ir + self._voxel_map.shape[2] * iphi + iz
+        else:
+            icell = self._voxel_map.shape[1] * ir + iz
+        isource = self.voxel_map_mv[icell]  # index of the light source in spectral array

```

Comment:
The `emission_function()` is needed only for compatibility with `NumericalIntegrator`, its performance is poor anyway (memory access is a bottleneck). However, now it's decorated with the same Cython directives as the `integrate()` function of Ray Transfer Integrators.

When these emitters are used with Ray Transfer Objects, it is guaranteed that indices are within bounds because Ray Transfer Objects work as bounding boxes. However, I didn't notice any performance gain from `@cython.boundscheck(False)`, so I removed this directive from the `integrate()` and `emission_function()` for safety.